### PR TITLE
[NL] add text descriptions to categories

### PR DESCRIPTION
**Pull request by beets**
It sounds very scripted and not too happy with how the logic is split. Was torn between putting it fully in each fulfiller or while building the page config. We might have to come up with a format string that each fulfiller uses, providing most of the string, which is filled in with sv and place names, etc, while building page config.

Demo scripts in action:
[latest](https://user-images.githubusercontent.com/6052978/216557934-ce87a394-e71b-4083-ae68-7556f4d72b7e.png)
[script 2](https://user-images.githubusercontent.com/6052978/216557945-ee2af5ff-6348-4d98-b434-f063d3d0d15d.png)
[script 1](https://user-images.githubusercontent.com/6052978/216557959-203d8777-efc8-46ec-a205-4b62ad183e7d.png)

Newer screenshots:
[latest](https://user-images.githubusercontent.com/6052978/216723164-5ea8275a-cdfa-4daf-a492-195ab3707419.png)
[script 1](https://user-images.githubusercontent.com/6052978/216723230-692be430-31c6-4c14-a06a-3badeed8955d.png)
[script 2](https://user-images.githubusercontent.com/6052978/216723374-cc94740a-1711-4d73-b22d-cf3065424e50.png)

### CODE REVIEW

**Comment by pradh** [comment]
> It sounds very scripted and not too happy with how the logic is split. Was torn between putting it fully in each fulfiller or while building the page config. We might have to come up with a format string that each fulfiller uses, providing most of the string, which is filled in with sv and place names, etc, while building page config.
> 
> Demo scripts in action: [latest](https://user-images.githubusercontent.com/6052978/216557934-ce87a394-e71b-4083-ae68-7556f4d72b7e.png) [script 2](https://user-images.githubusercontent.com/6052978/216557945-ee2af5ff-6348-4d98-b434-f063d3d0d15d.png) [script 1](https://user-images.githubusercontent.com/6052978/216557959-203d8777-efc8-46ec-a205-4b62ad183e7d.png)

This feels like a nice start to me.

Approach wise... the fulfiller doesn't have a full view of all the charts to be added, whereas the page config builder does (e.g., it can make global decisions based on number of ranked charts within a block, etc).  Feels slightly better to build description strings in page config (there is per chart-type handling there)?  But the fulfiller should put whatever additional info is needed in ChartSpec.attr.

nit:  since this is going to grow, perhaps all description functions can be in its own file instead of page_config_next?

**Comment by shifucun** [comment]
> > It sounds very scripted and not too happy with how the logic is split. Was torn between putting it fully in each fulfiller or while building the page config. We might have to come up with a format string that each fulfiller uses, providing most of the string, which is filled in with sv and place names, etc, while building page config.
> > Demo scripts in action: [latest](https://user-images.githubusercontent.com/6052978/216557934-ce87a394-e71b-4083-ae68-7556f4d72b7e.png) [script 2](https://user-images.githubusercontent.com/6052978/216557945-ee2af5ff-6348-4d98-b434-f063d3d0d15d.png) [script 1](https://user-images.githubusercontent.com/6052978/216557959-203d8777-efc8-46ec-a205-4b62ad183e7d.png)
> 
> This feels like a nice start to me.
> 
> Approach wise... the fulfiller doesn't have a full view of all the charts to be added, whereas the page config builder does (e.g., it can make global decisions based on number of ranked charts within a block, etc). Feels slightly better to build description strings in page config (there is per chart-type handling there)? But the fulfiller should put whatever additional info is needed in ChartSpec.attr.
> 
> nit: since this is going to grow, perhaps all description functions can be in its own file instead of page_config_next?

+1 on having a separate module that takes in an Utterance (with chart spec) and populate description for each chart.

**Comment by beets** [comment]
thanks for the review and suggestions - will pull this out into a separate module and ping back

**Comment by beets** [comment]
ptal! made a new module and simplified the code.

File: server/lib/nl/page_config_next.py [hunk 1]
```diff
@@ -82,6 +72,15 @@ def build_page_config(
   all_svs = list(all_svs)
   sv2name = utils.get_sv_name(all_svs)
 
+  # Add a human answer to the query
+  try:
+    desc = lib_desc.build_category_description(uttr, sv2name)
+    if desc:
+      category.description = desc
+  except Exception as err:
+    logging.warning("Error building category description", err)
```
Comment (pradh) [inline server/lib/nl/page_config_next.py:81]: nit: should we add `utils.update_counter(uttr.counters, 'failed_category_description_build', str(err))` or something?

File: server/lib/nl/fulfillment/base.py [hunk 2]
```diff
@@ -230,7 +235,7 @@ def overview_fallback(state: PopulateState, places: List[Place],
                          block_id=state.block_id,
                          include_percapita=False)
   return add_chart_to_utterance(ChartType.PLACE_OVERVIEW, state, chart_vars,
-                                places, chart_origin)
+                                places, chart_origin, "overview")
```
Comment (pradh) [inline server/lib/nl/fulfillment/base.py:238]: remove last arg?

File: server/lib/nl/fulfillment/base.py [hunk 3]
```diff
@@ -60,9 +62,12 @@ class ChartVars:
...
 #
-def add_chart_to_utterance(chart_type: ChartType, state: PopulateState,
-                           chart_vars: ChartVars, places: List[Place],
-                           primary_vs_secondary: ChartOriginType) -> bool:
+def add_chart_to_utterance(chart_type: ChartType,
+                           state: PopulateState,
+                           chart_vars: ChartVars,
+                           places: List[Place],
+                           primary_vs_secondary: ChartOriginType,
+                           desc: str = None) -> bool:
```
[the remark this answers is not in the document]
Comment (beets) [reply server/lib/nl/fulfillment/base.py]: dropped description

File: server/lib/nl/fulfillment/base.py [hunk 4]
```diff
@@ -222,6 +228,7 @@ def _build_chart_vars(state: PopulateState, sv: str,
   if utils.is_topic(sv):
     topic_vars = topic.get_topic_vars(sv, rank)
     peer_groups = topic.get_topic_peers(topic_vars)
+    state.uttr.topic = sv
```
[the remark this answers is not in the document]
Comment (beets) [reply server/lib/nl/fulfillment/base.py]: yes - good point. moved it to topic and feels cleaner too

File: server/lib/nl/page_config_next.py [hunk 5]
```diff
@@ -94,6 +182,7 @@ def build_page_config(
     block_id = cspec.attr['block_id']
     if block_id != prev_block_id:
       if block:
+        continue
```
[the remark this answers is not in the document]
Comment (beets) [reply server/lib/nl/page_config_next.py]: this was for dev -- removed

File: server/lib/nl/fulfillment/simple.py [hunk 6]
```diff
@@ -37,6 +40,9 @@ def populate(uttr: Utterance) -> bool:
 def _populate_cb(state: PopulateState, chart_vars: ChartVars,
                  places: List[Place], chart_origin: ChartOriginType) -> bool:
   logging.info('populate_cb for simple')
+  description = TIMELINE_TEXT.format(place_name=", ".join(
+      [p.name for p in places]),
+                                     svs=", ".join(chart_vars.svs))
```
[the remark this answers is not in the document]
Comment (beets) [reply server/lib/nl/fulfillment/simple.py]: not anymore. dropped

**Review by pradh — APPROVED** [review]
Skimmed through the changes.  Looking good.  We can get it in and improve :)

File: server/lib/nl/page_config_next.py [hunk 7]
```diff
@@ -53,6 +59,82 @@
...
+      desc = _EVENT_TYPE_TO_DISPLAY_NAME[event_types[0]].lower()
+  elif uttr.topic:
+    desc = topic.get_topic_name(uttr.topic).lower()
+
+  by = ''
+  if len(uttr.classifications) and isinstance(
+      uttr.classifications[0].attributes, ContainedInClassificationAttributes):
+    place_type = uttr.classifications[0].attributes.contained_in_place_type
+    by = ' by ' + place_type.value.lower()
+
```
[the remark this answers is not in the document]
Comment (beets) [reply server/lib/nl/page_config_next.py]: good point - added it as a todo

File: server/lib/nl/page_config_next.py (hunk 1 above)
Comment (beets) [reply server/lib/nl/page_config_next.py:82]: done

File: server/lib/nl/fulfillment/base.py (hunk 2 above)
Comment (beets) [reply server/lib/nl/fulfillment/base.py]: dropped -thanks for catching that

**Comment by beets** [comment]
thanks for the review!

**Pull request merged**